Pathlib support

clbarnes opened the pull request:
Minor changes to utility functions to allow convenience functions to work with `os.PathLike`s; see #261 and #601. These now just normalise to `str`, which the DirectoryStore was doing explicitly before anyway. It might be nice to avoid that and for the `path` member of things containing a path to store (/return) it as a `Path`, but that was a more invasive change; the main value here is avoiding the errors.

One module defined a `Path` newtype for typing purposes; I renamed that to `PathType` for clarity.

I found it a little confusing to see "path" used to describe zarr's internal keys; looking through the code, you have to get whether it's talking about key or e.g. file system path from context. Would changing that verbiage be considered in an upcoming major version increment? Another "going forward" thought: it's a bit of a pain/ source of errors to manipulate path-like things as strings, which is why pathlib exists. Would you consider something like

```python
from pathlib import PurePosixPath

class ZarrKey(PurePosixPath):
    pass
```

for internal use to begin with? On the flip side, we'd need to be more careful about `os.PathLike` checks, but there are only a couple in there.

Lastly, where in `release.rst` do new features go? There doesn't seem to be an entry for "current release" features.

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in docs/tutorial.rst
    - Nothing to document as I would assume this behaviour from anything interacting with the file system
* [x] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

joshmoore commented:
> Would you consider something like ... `class ZarrKey(PurePosixPath)` for internal use to begin with?

Interesting. I'd be especially happy if I could attach extra metadata to that ;) see https://github.com/zarr-developers/zarr-python/issues/769#issuecomment-861719216

> Lastly, where in `release.rst` do new features go?

Just add to the top. (The mainline may now have an "Unreleased" section at the top)

clbarnes commented:
Added documentation to release.rst. Build failures are (I hope) unrelated; a segfault coming from numpy.

clbarnes commented:
Rebased and green!

joshmoore commented on zarr/storage.py:
```diff
@@ -76,7 +76,7 @@
 
 
-Path = Union[str, bytes, None]
+PathType = Union[str, bytes, None]
```
This is probably the only change that I don't feel competent to individually review.

Does anyone know the possible impact of this? In my experience, this would count as a breaking change.

clbarnes replied:
This could be reverted. From its usage it looked pretty internal (and only in the one module) but I suppose someone downstream could be using it.

joshmoore replied:
> but I suppose someone downstream could be using it.

*nods* That was my thought. I have _zero_ evidence that someone is actually doing that, but just in terms of the semvers'ioning

joshmoore reviewed (commented):
In general, looks good from my side. I've noted one renaming issue on which I'd appreciate thoughts from others.

clbarnes commented:
Now rebased on master; I've also reverted the `Path` rename and have used `import pathlib; ... pathlib.Path` everywhere it's used to be maximally explicit.

joshmoore reviewed (approved):
Thanks for the rebase, @clbarnes! To my eye, this is now a very _non_-invasive widening of what the open methods will accept.
